### PR TITLE
feat(compiler): add support for external templates

### PR DESCRIPTION
**Pull request by tripodsan**
first draft to implement #217

### CODE REVIEW

**Comment by jantimon** [comment]
This looks really cool 👍

Just one small thing the JSPureTemplate is has `$scriptId` which isn't used at all:

https://github.com/adobe/htlengine/blob/a463b9569f022104554fca043bb75e4dfbf8246f/src/compiler/JSPureTemplate.js#L14-L16

**Comment by tripodsan** [comment]
> One bug I found is if a template file is using itself:

this is indeed a problem... I'll see if we can fix it here or if we should create a follow up bug.

**Comment by tripodsan** [comment]
> Just one small thing the `JSPureTemplate` has `$scriptId` which isn't used at all:

since it's not needed anymore, we could also use the default code template again....

File: src/compiler/JSCodeGenVisitor.js [hunk 1]
```diff
@@ -265,6 +276,10 @@ module.exports = class JSCodeGenVisitor {
       this._dom.doctype(cmd);
     } else if (cmd instanceof Comment) {
       this._dom.comment(cmd);
+    } else if (cmd instanceof ExternalCode) {
+      this.pushBlock('foo');
```
Comment (jantimon) [inline src/compiler/JSCodeGenVisitor.js:280]: What does this do? 🤔

File: src/compiler/JSPureTemplate.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    col: runtime.col,
+    exec: runtime.exec.bind(runtime),
+    xss: runtime.xss.bind(runtime),
+    listInfo: runtime.listInfo.bind(runtime),
+    use: runtime.use.bind(runtime),
+    slyResource: runtime.resource.bind(runtime),
+    call: runtime.call.bind(runtime),
+    template: runtime.template.bind(runtime),
+    dom: runtime.dom,
+  };
```
Comment (jantimon) [inline src/compiler/JSPureTemplate.js:26]: would it be possible to replace this part with `const $ = runtime` or `const $ = runtime.templateHelpers()`?

Comment (tripodsan) [reply src/compiler/JSPureTemplate.js:26]: I guess :-)

File: src/compiler/JSCodeGenVisitor.js (hunk 1 above)
Comment (tripodsan) [reply src/compiler/JSCodeGenVisitor.js:280]: it opens a new template block with the name 'foo'. maybe the name can be left empty.

File: src/compiler/JSPureTemplate.js (hunk 2 above)
Comment (tripodsan) [reply src/compiler/JSPureTemplate.js:26]: let's tackle this in a separate PR

File: test/multi/templates/header.htl [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<template data-sly-template.heading="${@ data}">
+    <h1>${data.title}</h1>
+    <sly data-sly-call="${separator}"/>
+</template>
+<template data-sly-template.separator>--------------------------------------</template>
```
Comment (tripodsan) [inline test/multi/templates/header.htl:5]: @jantimon Here, I added a test for a template that uses a template that is later defined in the same script. and it works now :-)

Comment (jantimon) [reply test/multi/templates/header.htl:5]: Really cool 👍

**Pull request merged**